src/achievement-rule.js

```diff
@@ -0,0 +1,214 @@
+module.exports = class AchievementRule {
+  constructor(rule, name) {
+    this.name = name;
+    const requirements = rule['requirements'] === undefined ? [] : rule['requirements'];
+    this.requirements = requirements.map(requirement => Requirement.fromYamlRequirement(requirement));
+    this.replaces = rule['replaces'] === undefined ? [] : rule['replaces'] ;
+
+    let maxAwarded = rule['maxAwarded'];
+    let maxAwardedTotal = rule['maxAwardedTotal'];
+    if (maxAwarded === undefined && maxAwardedTotal === undefined) {
+      // By default, an achievement can only be awarded once, ever.
+      maxAwarded = maxAwardedTotal = 1;
+    } else if (maxAwarded === undefined && maxAwardedTotal !== undefined) {
+      // If only maxAwardedTotal is set, we can also safely set maxAwarded to
+      // the same value.
+      maxAwarded = maxAwardedTotal;
+    } else if (maxAwarded !== undefined && maxAwardedTotal === undefined) {
+      // In case only maxAwarded is set, we need to set maxAwardedTotal to +Infinity.
+      maxAwardedTotal = Number.POSITIVE_INFINITY;
+    }
+    if (maxAwarded > maxAwardedTotal) {
+      throw new Error(`The achievement "${this.name}" has maxAwarded > maxAwardedTotal (${this.maxAwarded} > ${this.maxAwardedTotal}).`);
+    }
+    this.maxAwarded = maxAwarded;
+    this.maxAwardedTotal = maxAwardedTotal;
```

Comment:
I moved these classes into a separate file to make the `AchievementRule` easier to test. Apart from the `maxAwarded<Total>` logic above and around line 50, I didn't change anything.